samtranslator/utils/traverse.py

```diff
@@ -0,0 +1,104 @@
+from typing import Any, Dict, List
+
+from samtranslator.utils.resolve_actions import ResolveAction
+
+
+def traverse(template: Dict[str, Any], actions: List[ResolveAction]) -> Any:
+    """
+    Calls the correct traversal method
+
+    :param template: The template that needs modifying
+    :param actions: The actions that need to be performed to the template
+    :return: Modified Template
+
+    TODO: Add multiple strategies and choose between those with an Optional value
+    """
+    return _preorder_traverse(template, actions)
+
+
+def _execute_actions(template: Dict[str, Any], actions: List[ResolveAction]) -> Any:
+    """
+    Calls the resolver methods when they are in the list of ResolveActions. Then calls the actions passed to traverse.
+
+    :param template: The template that needs modifying
+    :param actions: actions to be executed.
+    :return: Modified Template
+    """
+    if len(actions) == 0:
+        return template
+
+    for action in actions:
+        action.execute(template)
+
+    return template
+
+
+def _preorder_traverse(
```

Comment:
this and `_traverse_dict/_traverse_list` could be in abstract class so that they can be reused.

Reply:
Will leave to a later refactor due to differences in input values.